### PR TITLE
Allow unused destructors to be completely elided

### PR DESCRIPTION
**Pull request by sbc100**
Depends on https://github.com/llvm/llvm-project/pull/68758

Fixes: #19993

### CODE REVIEW

File: test/test_other.py [hunk 1]
```diff
@@ -14048,3 +14048,8 @@ def test_emmalloc_in_addition(self, args):
     emmalloc = path_from_root('system', 'lib', 'emmalloc.c')
     self.run_process([EMCC, test_file('other/test_emmalloc_in_addition.c'), emmalloc] + args)
     self.assertContained('success', self.run_js('a.out.js'))
+
+  def test_unused_destructor(self):
+    self.do_runf(test_file('other/test_unused_destructor.c'), emcc_args=['-Wl,-save-temps', '-flto', '-O2'])
+    # Verify that the string constant in the desructor is not included in the binary
```
Comment (kripken) [inline test/test_other.py:14054]:
```suggestion
    # Verify that the string constant in the destructor is not included in the binary
```

File: tools/building.py [hunk 2]
```diff
@@ -178,6 +178,15 @@ def lld_flags_for_executable(external_symbols):
   if settings.LINKABLE:
     cmd.append('--export-dynamic')
 
+  if settings.LTO and not settings.EXIT_RUNTIME:
+    # The WebAssembly backend can generate new references to `__cxa_atexit` at
+    # LTO time.  This `-u` flag forces the `__cxa_atexit` symbol to be
+    # included at LTO time.  For other such symbols we exclude them from LTO
+    # and always build them as normal object files, but that would inhinbit the
```
Comment (kripken) [inline tools/building.py:185]:
```suggestion
    # and always build them as normal object files, but that would inhibit the
```

File: tools/building.py [hunk 3]
```diff
@@ -178,6 +178,15 @@ def lld_flags_for_executable(external_symbols):
...
+    # LowerGlobalDtors optimization which allows destrutors to be completely
```
Comment (kripken) [inline tools/building.py:186]:
```suggestion
    # LowerGlobalDtors optimization which allows destructors to be completely
```

File: test/test_other.py [hunk 4]
```diff
@@ -14048,3 +14048,8 @@ def test_emmalloc_in_addition(self, args):
     emmalloc = path_from_root('system', 'lib', 'emmalloc.c')
     self.run_process([EMCC, test_file('other/test_emmalloc_in_addition.c'), emmalloc] + args)
     self.assertContained('success', self.run_js('a.out.js'))
+
+  def test_unused_destructor(self):
+    self.do_runf(test_file('other/test_unused_destructor.c'), emcc_args=['-Wl,-save-temps', '-flto', '-O2'])
```
Comment (kripken) [inline test/test_other.py:14053]: Why do we need save-temps here?

**Review by kripken — COMMENTED** [review]
Great!

File: test/test_other.py (hunk 4 above)
Comment (sbc100) [reply test/test_other.py:14053]: Nope. Removed.

**Review by kripken — APPROVED** [review]

**Pull request merged**